functions/utils/recording_events.py

```diff
@@ -0,0 +1,245 @@
+from os import getenv as env
+from dotenv import load_dotenv
+from os.path import join, dirname
+from .common import TIMESTAMP_FORMAT
+from pytz import utc
+from itertools import chain
+
+logger = logging.getLogger()
+
+load_dotenv(join(dirname(__file__), "../.env"))
+
+RECORDING_EVENTS_TABLE = env("RECORDING_EVENTS_TABLE", None)
+
+
+RECORDING_EVENTS = [
+    "recording.started",
+    "recording.paused",
+    "recording.resumed",
+    "recording.stopped",
+]
+
+START_EVENTS = ["recording.started", "recording.resumed"]
+END_EVENTS = ["recording.paused", "recording.stopped"]
+
+# This was copied from status.py:
+# If RECORDING_EVENTS_TABLE is not set we assume the events table
+# is not being used. (For example tasks.py imports utils but doesn't
+# use the recording times table)
+recording_events_table = None
+if RECORDING_EVENTS_TABLE:
+    dynamodb = boto3.resource("dynamodb")
+    recording_events_table = dynamodb.Table(RECORDING_EVENTS_TABLE)
+
+
+def get_recording_events(zoom_uuid):
+    """
+    Gets recording events from the database by zoom uuid.
+    """
+    r = recording_events_table.get_item(Key={"zoom_uuid": zoom_uuid})
+    logger.debug(f"Recording events from db: ${r}")
+    return r["Item"] if "Item" in r else None
+
+
+def set_recording_events(
+    zoom_uuid,
+    zoom_event,
+    zoom_event_timestamp,  # in ms
+    meeting_id=None,
+):
+    """
+    Records a recording event in the database. Events are stored by zoom uuid in a list of
+    {
+        "event": "ZOOM_EVENT",  # recording.started, paused, resumed, stopped
+        "timestamp": "TIMESTAMP_IN_SECS",
+    }
+    """
+    # If not one of desired events or no timestamp, ignore
+    if zoom_event not in RECORDING_EVENTS or not zoom_event_timestamp:
+        return
+
+    # If this is an event for recording started/paused/resumed/stopped,
+    # record the event time in the table in a list.
+    formatted_local_time = datetime.strftime(
+        datetime.fromtimestamp(int(zoom_event_timestamp / 1000)),
+        "%m/%d/%Y %H:%M:%S",
+    )
+    logger.info(
+        f"Record event for '{zoom_uuid}': {zoom_event} at {formatted_local_time}"
+    )
+
+    try:
+        new_recording_event_list = [
+            {
+                "event": zoom_event,
+                "timestamp": int(zoom_event_timestamp / 1000),  # in secs
+            }
+        ]
+        utcnow = datetime.utcnow()
+        update_expression = (
+            "SET "
+            "expiration = :expiration, "
+            "recording_event_times = list_append("
+            "if_not_exists(recording_event_times, :empty_list), "
+            ":new_recording_event"
+            ")"
+        )
+        # Expiration 120 days, which should cover at least a semester
+        expression_attribute_values = {
+            ":expiration": int((utcnow + timedelta(days=120)).timestamp()),
+            ":new_recording_event": new_recording_event_list,
+            ":empty_list": [],
+        }
+
+        if meeting_id:
```

Comment:
When would there not be a `meeting_id`? Also why is it included in the item? I don't see it being referenced in queries.

Reply:
It isn't. I can remove it.